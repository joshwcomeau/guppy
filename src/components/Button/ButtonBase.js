// @flow
import React, { Component } from 'react';
import styled from 'styled-components';

import { RAW_COLORS, COLORS } from '../../constants';

type Size = 'xsmall' | 'small' | 'medium' | 'large';

type Props = {
  size: Size,
  background?: string,
  hoverBackground?: string,
  textColor?: string,
  noPadding?: boolean,
  activeSplat?: boolean,
  disabled?: boolean,
  children: React$Node,
};

class ButtonBase extends Component<Props> {
  static defaultProps = {
    size: 'medium',
    background: RAW_COLORS.gray[200],
    textColor: COLORS.text,
  };

  getButtonElem = (size: Size) => {
    switch (size) {
      case 'xsmall':
        return XSmallButton;
      case 'small':
        return SmallButton;
      default:
      case 'medium':
        return MediumButton;
      case 'large':
        return LargeButton;
    }
  };

  render() {
    const {
      size,
      children,
      background,
      hoverBackground,
      textColor,
      disabled,
      ...delegated
    } = this.props;

    const Elem = this.getButtonElem(size);

    return (
      <Elem
        background={background}
        hoverBackground={hoverBackground}
        textColor={textColor}
        disabled={disabled}
        {...delegated}
      >
        {children}
      </Elem>
    );
  }
}

const ButtonBaseStyles = styled.button`
  position: relative;
  border: 0;
  display: inline-flex;
  justify-content: center;
  align-items: center;
  background: ${props => props.background};
  color: ${props => props.textColor};
  cursor: pointer;
  outline: none;
  white-space: nowrap;

  &:hover {
    background: ${props => props.hoverBackground};
  }

  &:disabled {
    filter: grayscale(100%);
    opacity: 0.75;
    cursor: initial;
  }

  &:not(:disabled):active {
    transform-origin: center center;
    transform: ${props => props.activeSplat && 'scale(1.1)'};
  }
`;

export const XSmallButton = styled(ButtonBaseStyles)`
  padding: ${props => (props.noPadding ? '0px' : '0px 12px')};
  height: ${props => (props.noPadding ? 'auto' : '22px')};
  border-radius: 15px;
  font-size: 12px;
`;

export const SmallButton = styled(ButtonBaseStyles)`
  padding: ${props => (props.noPadding ? '0px' : '0px 14px')};
  height: ${props => (props.noPadding ? 'auto' : '30px')};
  border-radius: 17px;
  font-size: 14px;
`;

export const MediumButton = styled(ButtonBaseStyles)`
  padding: ${props => (props.noPadding ? '0px' : '0px 20px')};
  height: ${props => (props.noPadding ? 'auto' : '38px')};
  border-radius: 19px;
  font-size: 16px;
`;

export const LargeButton = styled(ButtonBaseStyles)`
  padding: ${props => (props.noPadding ? '0px' : '0 32px')};
  height: ${props => (props.noPadding ? 'auto' : '48px')};
  border-radius: 24px;
  font-size: 24px;
`;

export default ButtonBase;
