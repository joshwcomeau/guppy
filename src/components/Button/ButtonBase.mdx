---
name: ButtonBase
menu: Reusable Components
---

import { Playground, Props } from 'docz';
import ButtonBase from './ButtonBase';

# ButtonBase

> Note: You probably don't want to use this component directly. It's just the
> underlying shared bit of code between FillButton and StrokeButton.

<Props of={ButtonBase} />

## Basic usage

<Playground>
  <ButtonBase>Default</ButtonBase>
</Playground>
