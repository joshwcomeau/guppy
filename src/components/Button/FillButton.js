// @flow
import React, { Component } from 'react';

import { COLORS, GRADIENTS } from '../../constants';

import ButtonBase from './ButtonBase';

type Props = {
  colors?: Array<string>,
  hoverColors?: Array<string>,
  textColor: string,
  children: React$Node,
  disabled?: boolean,
};

export const wrapColorsInGradient = (colors?: Array<string> | string) => {
  if (!Array.isArray(colors)) {
    return colors;
  }

  if (colors.length === 1) {
    return colors[0];
  }

  return `linear-gradient(
    45deg,
    ${colors.join(',')}
  )`;
};

class FillButton extends Component<Props> {
  static defaultProps = {
    colors: GRADIENTS.primary,
    textColor: COLORS.textOnBackground,
  };

  render() {
    const { colors, hoverColors, children, ...delegated } = this.props;

    const background = wrapColorsInGradient(colors);
    const hoverBackground = wrapColorsInGradient(hoverColors) || background;

    return (
      <ButtonBase
        activeSplat
        background={background}
        hoverBackground={hoverBackground}
        {...delegated}
      >
        {children}
      </ButtonBase>
    );
  }
}

export default FillButton;
