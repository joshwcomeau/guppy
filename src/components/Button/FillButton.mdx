---
name: FillButton
menu: Reusable Components
---

import { Playground, Props } from 'docz';

import { RAW_COLORS } from '../../constants';
import FillButton from './FillButton';

# FillButton

<Props of={FillButton} />

## Basic usage

<Playground>
  <FillButton>Default</FillButton>
</Playground>

## Sizes

<Playground>
  <FillButton size="xsmall">
    Extra-small button
  </FillButton>
</Playground>

<Playground>
  <FillButton size="small">
    Small button
  </FillButton>
</Playground>

<Playground>
  <FillButton size="medium">
    Medium button
  </FillButton>
</Playground>

<Playground>
  <FillButton size="large">
    Large button
  </FillButton>
</Playground>

## Colors

<Playground>
  <FillButton colors={[RAW_COLORS.green[500], RAW_COLORS.green[500]]}>
    Solid green
  </FillButton>
</Playground>

<Playground>
  <FillButton colors={[RAW_COLORS.orange[500], RAW_COLORS.pink[300]]}>
    Orange to pink
  </FillButton>
</Playground>

<Playground>
  <FillButton colors={[RAW_COLORS.teal[700], RAW_COLORS.lime[500]]}>
    Teal to lime
  </FillButton>
</Playground>

## Disabled

<Playground>
  <FillButton disabled>
  FillButton
  </FillButton>
</Playground>
