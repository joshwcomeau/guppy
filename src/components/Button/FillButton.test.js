/* eslint-disable flowtype/require-valid-file-annotation */
import React from 'react';
import { mount } from 'enzyme';
import FillButton, { wrapColorsInGradient } from './FillButton';
import ButtonBase from './ButtonBase';

describe('FillButton component', () => {
  it('should render button filled', () => {
    const wrapper = mount(<FillButton />);
    expect(wrapper.find(ButtonBase)).toMatchSnapshot();
  });

  describe('Wrap colors in gradient', () => {
    const testColor = '#fff';
    it('should return single color (string)', () => {
      expect(wrapColorsInGradient(testColor)).toEqual(testColor);
    });
    it('should return single color (array)', () => {
      expect(wrapColorsInGradient([testColor])).toEqual(testColor);
    });

    it('should return a gradient', () => {
      const colors = [testColor, '#000'];
      expect(wrapColorsInGradient(colors)).toEqual(`linear-gradient(
    45deg,
    ${colors.join(',')}
  )`);
    });
  });
});
