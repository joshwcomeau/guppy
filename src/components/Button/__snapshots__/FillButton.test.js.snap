// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FillButton component should render button filled 1`] = `
.c1 {
  position: relative;
  border: 0;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: linear-gradient( 45deg, #651fff,#D500F9 );
  color: #FFF;
  cursor: pointer;
  outline: none;
  white-space: nowrap;
}

.c1:hover {
  background: linear-gradient( 45deg, #651fff,#D500F9 );
}

.c1:disabled {
  -webkit-filter: grayscale(100%);
  filter: grayscale(100%);
  opacity: 0.75;
  cursor: initial;
}

.c1:not(:disabled):active {
  -webkit-transform-origin: center center;
  -ms-transform-origin: center center;
  transform-origin: center center;
  -webkit-transform: scale(1.1);
  -ms-transform: scale(1.1);
  transform: scale(1.1);
}

.c0 {
  padding: 0px 20px;
  height: 38px;
  border-radius: 19px;
  font-size: 16px;
}

<ButtonBase
  activeSplat={true}
  background="linear-gradient(
    45deg,
    #651fff,#D500F9
  )"
  hoverBackground="linear-gradient(
    45deg,
    #651fff,#D500F9
  )"
  size="medium"
  textColor="#FFF"
>
  <Styled(styled.button)
    activeSplat={true}
    background="linear-gradient(
    45deg,
    #651fff,#D500F9
  )"
    hoverBackground="linear-gradient(
    45deg,
    #651fff,#D500F9
  )"
    textColor="#FFF"
  >
    <styled.button
      activeSplat={true}
      background="linear-gradient(
    45deg,
    #651fff,#D500F9
  )"
      className="c0"
      hoverBackground="linear-gradient(
    45deg,
    #651fff,#D500F9
  )"
      textColor="#FFF"
    >
      <button
        className="c0 c1"
      />
    </styled.button>
  </Styled(styled.button)>
</ButtonBase>
`;
